Rubberduck.Parsing/Symbols/ModuleBodyElementDeclaration.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Antlr4.Runtime;
+using Rubberduck.Parsing.Annotations;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Parsing.Symbols
+{
+    public abstract class ModuleBodyElementDeclaration : Declaration, IParameterizedDeclaration, ICanBeInterfaceMember, ICanBeDefaultMember
+    {
+        protected ModuleBodyElementDeclaration(
+            QualifiedMemberName name,
+            Declaration parent,
+            Declaration parentScope,
+            string asTypeName,
+            VBAParser.AsTypeClauseContext asTypeContext,
+            string typeHint,
+            Accessibility accessibility,
+            DeclarationType type,
+            ParserRuleContext context,
+            ParserRuleContext attributesPassContext,
+            Selection selection,
+            bool isArray,
+            bool isUserDefined,
+            IEnumerable<IAnnotation> annotations,
+            Attributes attributes)
+            : base(
+                name,
+                parent,
+                parentScope,
+                asTypeName,
+                typeHint,
+                false,
+                false,
+                accessibility,
+                type,
+                context,
+                attributesPassContext,
+                selection,
+                isArray,
+                asTypeContext,
+                isUserDefined,
+                annotations,
+                attributes)
+        {
+            if (parent is ClassModuleDeclaration classModule)
+            {
+                classModule.AddMember(this);
+            }
+        }
+   
+        public bool IsInterfaceImplementation => (_implementsResolved ? _implements : InterfaceImplemented) != null;
+
+        private bool _implementsResolved;
+        private ClassModuleDeclaration _implements;
+        public ClassModuleDeclaration InterfaceImplemented
```

Comment:
I hope this does not get called concurrently because it is not thread-safe. Maybe use a `Lazy<T>` instead of the flag `_implementsResolved`.

Reply:
Didn't even occur to me that this would need to be thread safe. `Lazy<T>` would be a much cleaner implementation anyway. Fix coming.